src/api/vst3.rs

```diff
@@ -0,0 +1,757 @@
+                        }
+                        // todo: error condition
+                    }
+                }
+                // drain parameter changes
+                // todo: interleave with events for sample accurate automation
+                if let Some(input_param_changes) = data.input_param_changes.upgrade() {
+                    for i in 0..input_param_changes.get_parameter_count() {
+                        if let Some(queue) = input_param_changes.get_parameter_data(i).upgrade() {
+                            let id = queue.get_parameter_id();
+                            for p in 0..queue.get_point_count() {
+                                let mut offset = 0;
+                                let mut value = 0.0;
+                                if queue.get_point(p, &mut offset as *mut _, &mut value as *mut _)
+                                    != 0
+                                {
+                                    break;
+                                }
+                                if let Some(param) = baseplug::api::vst3::param_for_vst3_id::<
+                                    <$plugin as Plugin>::Model,
+                                >(id)
+                                {
+                                    self.plugin().set_parameter(param, value as f32);
+                                }
+                            }
+                        }
+                    }
+                }
+                if data.inputs.is_null()
+                    || data.outputs.is_null()
+                    || data.num_inputs != data.num_outputs
+                    || (*data.inputs).buffers.is_null()
+                    || (*data.outputs).buffers.is_null()
+                {
+                    return 0;
+                }
+                let buffer_size = data.num_samples as usize;
+                if buffer_size == 0 {
+                    return 0;
+                }
+                let inputs = {
+                    let buffers = (&*data.inputs).buffers;
+                    [
+                        std::slice::from_raw_parts(buffers.offset(0) as *mut f32, buffer_size),
+                        std::slice::from_raw_parts(buffers.offset(1) as *mut f32, buffer_size),
+                    ]
+                };
+                let outputs = {
+                    let buffers = (&mut *data.outputs).buffers;
+                    [
+                        std::slice::from_raw_parts_mut(buffers.offset(0) as *mut f32, buffer_size),
+                        std::slice::from_raw_parts_mut(buffers.offset(1) as *mut f32, buffer_size),
+                    ]
+                };
+                let time = {
+                    if data.context.is_null() {
+                        // todo: figure this out
+                        baseplug::MusicalTime {
+                            bpm: 0.0,
+                            beat: 0.0,
+                        }
```

Comment:
should we pass in `MusicalTime` to the plugin as an `Option<MusicalTime>` to handle the case where the host doesn't provide any?